snippets/unflattenObject.md

```diff
@@ -5,29 +5,27 @@ tags: object,advanced
 
 Unflatten an object with the paths for keys.
 
-- Use `Object.keys(obj)` combined with `Array.prototype.reduce()` to convert flattened path node to a leaf node.
-- If the value of a key contains a dot delimiter (`.`), use `Array.prototype.split('.')`, string transformations and `JSON.parse()` to create an object, then `Object.assign()` to create the leaf node.
-- Otherwise, add the appropriate key-value pair to the accumulator object.
+- Iterate over the object keys using `Object.keys(obj)` and `Array.prototype.forEach()`.
+- Split each key with a dot delimiter (`.`) using `Array.prototype.split('.')`and use `Array.prototype.reduce()` to convert flattened paths to leaf node.
+- If the current accumulator already contains value against a particular key, return its value as the next accumulator.
+- Otherwise, add the appropriate key-value pair to the accumulator object and return value as the accumulator.
+
 
 ```js
-const unflattenObject = obj =>
-  Object.keys(obj).reduce((acc, k) => {
-    if (k.indexOf('.') !== -1) {
-      const keys = k.split('.');
-      Object.assign(
-        acc,
-        JSON.parse(
-          '{' +
-            keys.map((v, i) => (i !== keys.length - 1 ? `"${v}":{` : `"${v}":`)).join('') +
-            obj[k] +
-            '}'.repeat(keys.length)
-        )
-      );
-    } else acc[k] = obj[k];
-    return acc;
-  }, {});
+const unflattenObject = obj =>{
+    var result = {}
+    Object.keys(obj).forEach((k)=>{
+      var keys = k.split('.')
+      keys.reduce(function(acc, e, i) {
+        return acc[e] || (acc[e] = isNaN(Number(keys[i + 1])) ? (keys.length - 1 === i ? obj[k] : {}) : [])
```

Comment:
Why the assignment to `acc[e]` in `acc[e] = isNaN(Number(keys[i + 1])`? What am I missing?

Reply:
@Chalarangelo  the assignmet to acc[e] basically adds the necessary nested key value pairs to the accumulator (which is a referenece to a value against the parents key). 
for example, say while resolving `{ 'a.b': 2, a.d: 3 }`.

Initially `acc[e]` is `{ }` so `acc['a']` becomes `{ }` after the assignment. Now, after assignment, the reference to `acc['a']` which is `{}`, is passed to the next call as accumulator.
Now for the next call with `b`, similarly when an assignment is again made to `acc['b']` it goes directly inside `acc['a']` because in this case the accumulator was a reference to `acc['a']'s` values.

I hope this explains.

Reply:
Yeah, it makes sense. Maybe we could unnest this assignment into its own line? It would probably be more readable. What do you think?

Reply:
Yes, I think that would be possible. But we will have to add an extra if statement. Shall I do this?